File: players/shairport_player.py
# players/shairport_player.py
import asyncio
import os
from lxml import etree
import base64
import hashlib

PLAYER_CACHE_DIR = "/tmp/shairport_art_cache/"

transport_codes = [
    'prsm',
    'paus',
    'pend'
]

transport_state_map = {'prsm': 'playing', 'pend': 'stopped', 'paus': 'paused'}

metadata_codes = [
    'asar',
    'minm',
    'PICT',
    'mden',
    'mdst'
]

codes = transport_codes + metadata_codes


class ShairportPlayer:
    """An asyncio-native player that reads and parses the raw XML stream from the Shairport Sync metadata pipe."""

    def __init__(self, pipe_path, session_manager):
        self.name = "AirPlay"
        self._pipe_path = pipe_path
        self._session_manager = session_manager
        self._buffer = ""  # Buffer is now a string
        self._pipe_fd = None
        self._pipe_closed_event = asyncio.Event()
        self._staged_track_info = {}

    def _save_pict_data(self, songid, image_bytes):
        """Saves raw image bytes to a temporary file and returns its path."""
        try:
            os.makedirs(PLAYER_CACHE_DIR, exist_ok=True)
            filename = f"{hashlib.md5(songid.encode()).hexdigest()}.tmp"
            filepath = os.path.join(PLAYER_CACHE_DIR, filename)

            if os.path.exists(filepath):
                return f"file://{filepath}"

            with open(filepath, "wb") as f:
                f.write(image_bytes)

            return f"file://{filepath}"
        except Exception as e:
            print(f"SHAIRPORT: Failed to save PICT data to temp file: {e}")
            return None

    def _parse_item_xml(self, item_xml_str):
        """Parses a single <item>...</item> block and returns its code and value."""
        root = etree.fromstring(item_xml_str)
        code_hex = root.find('code').text
        code = bytes.fromhex(code_hex).decode(
            'utf-8', 'ignore') if code_hex else ''

        value = None
        data_node = root.find('data')
        if data_node is not None and data_node.text:
            encoding = data_node.get('encoding')
            content = data_node.text.strip()
            if content:
                value = base64.b64decode(
                    content) if encoding == 'base64' else content

        return code, value

    def _process_buffer(self):
        """Processes the buffer and calls session manager with complete metadata."""
        should_commit_metadata = False

        while True:
            start_tag = "<item>"
            end_tag = "</item>"
            start_index = self._buffer.find(start_tag)
            end_index = self._buffer.find(end_tag)


            if start_index != -1 and end_index > start_index:
                full_item_len = end_index + len(end_tag)
                item_xml = self._buffer[start_index:full_item_len]
                self._buffer = self._buffer[full_item_len:]
                try:
                    code, value = self._parse_item_xml(item_xml)
                    print(self._buffer)
                    if code not in codes:
                        continue

                    if code in transport_codes:
                        self._session_manager.update_transport_state(
                            self.name, transport_state_map[code])
                    elif code == 'mdst':
                        self._staged_track_info = {}
                    elif value is not None:
                        self._staged_track_info[code] = value
                        if 'minm' in self._staged_track_info and 'asar' in self._staged_track_info:
                            should_commit_metadata = True

                except Exception as e:
                    print(f"SHAIRPORT: Failed to parse item block: {e}")
                    print(f"SHAIRPORT: Malformed XML chunk was: {item_xml}")
            else:
                break

        if should_commit_metadata:
            print("SHAIRPORT: Title and artist received. Committing metadata.")
            artist = self._staged_track_info.get('asar', b'').decode('utf-8')
            title = self._staged_track_info.get('minm', b'').decode('utf-8')

            cover_url = None
            if 'PICT' in self._staged_track_info:
                songid = f"airplay-{artist}-{title}"
                file_uri = self._save_pict_data(
                    songid, self._staged_track_info['PICT'])
                cover_url = file_uri

            standardized_state = {
                "songid": f"airplay-{artist}-{title}",
                "title": title,
                "artist": artist,
                "cover_url": cover_url
            }

            self._session_manager.update_metadata(
                self.name, standardized_state)

    def _on_pipe_data(self):
        """Synchronous callback executed by the event loop when the pipe has data."""
        try:
            chunk = os.read(self._pipe_fd, 4096)
            if not chunk:
                self._pipe_closed_event.set()
            else:
                self._buffer += chunk.decode('utf-8', 'ignore')
                self._process_buffer()
        except (BlockingIOError, InterruptedError):
            pass  # Expected when no data is ready
        except Exception as e:
            print(f"SHAIRPORT: Error in read callback: {e}")
            self._pipe_closed_event.set()

    async def start(self):
        """Main asyncio task that opens the pipe and registers a non-blocking reader."""
        loop = asyncio.get_running_loop()
        while True:
            try:
                print(f"SHAIRPORT: Opening pipe at {self._pipe_path}...")
                self._pipe_fd = os.open(
                    self._pipe_path, os.O_RDONLY | os.O_NONBLOCK)
                loop.add_reader(self._pipe_fd, self._on_pipe_data)
                print("SHAIRPORT: Pipe reader registered. Waiting for events.")

                await self._pipe_closed_event.wait()
            except Exception as e:
                print(f"SHAIRPORT: Main loop error: {e}")
            finally:
                if self._pipe_fd:
                    loop.remove_reader(self._pipe_fd)
                    os.close(self._pipe_fd)
                    self._pipe_fd = None
                self._pipe_closed_event.clear()
                self._buffer = ""
                print("SHAIRPORT: Cleanup complete. Retrying in 5s.")
                await asyncio.sleep(5)

File: src/players/shairport_player.rs
use crate::{
    config::ShairportConfig,
    state::{Metadata, PlayerState, SessionManager},
};
use anyhow::{Context, Result};
use async_trait::async_trait;
use base64::{Engine, engine::general_purpose::STANDARD};
use std::{fs, os::unix::fs::OpenOptionsExt, path::PathBuf};
use tokio::io::AsyncReadExt;

use super::{Player, ShairportPlayer};

const TRANSPORT_CODES: &[&str] = &["prsm", "paus", "pend"];
const METADATA_CODES: &[&str] = &["asar", "minm", "PICT", "mden", "mdst"];

impl ShairportPlayer {
    pub fn new(config: ShairportConfig, session_manager: SessionManager) -> Self {
        Self {
            config,
            session_manager,
            buffer: String::new(),
            staged_track_info: Default::default(),
        }
    }
}

#[async_trait]
impl Player for ShairportPlayer {
    fn name(&self) -> &'static str {
        "shairport"
    }

    fn enabled(&self) -> bool {
        self.config.enabled
    }

    async fn start(self: Box<Self>) -> Result<()> {
        let mut player = *self;

        if !player.config.enabled {
            println!("SHAIRPORT: disabled.");
            return Ok(());
        }

        loop {
            if let Err(err) = player.run_once().await {
                println!("SHAIRPORT: Main loop error: {err:#}");
            }
            player.buffer.clear();
            println!(
                "SHAIRPORT: Cleanup complete. Retrying in {}s.",
                player.config.retry_secs
            );
            tokio::time::sleep(std::time::Duration::from_secs(player.config.retry_secs)).await;
        }
    }
}

impl ShairportPlayer {
    async fn run_once(&mut self) -> Result<()> {
        println!(
            "SHAIRPORT: Opening pipe at {}...",
            self.config.pipe_path.display()
        );
        let file = std::fs::OpenOptions::new()
            .read(true)
            .custom_flags(0o4000)
            .open(&self.config.pipe_path)
            .with_context(|| {
                format!("opening metadata pipe {}", self.config.pipe_path.display())
            })?;
        let mut file = tokio::fs::File::from_std(file);
        println!("SHAIRPORT: Pipe reader registered. Waiting for events.");

        let mut chunk = [0u8; 4096];
        loop {
            match file.read(&mut chunk).await {
                Ok(0) => break,
                Ok(n) => {
                    self.buffer.push_str(&String::from_utf8_lossy(&chunk[..n]));
                    self.process_buffer().await;
                }
                Err(err) if err.kind() == std::io::ErrorKind::WouldBlock => {
                    tokio::time::sleep(std::time::Duration::from_millis(100)).await;
                }
                Err(err) => return Err(err.into()),
            }
        }
        Ok(())
    }

    async fn process_buffer(&mut self) {
        let mut should_commit_metadata = false;

        while let Some(start_index) = self.buffer.find("<item>") {
            let Some(relative_end_index) = self.buffer[start_index..].find("</item>") else {
                break;
            };
            let full_item_len = start_index + relative_end_index + "</item>".len();
            let item_xml = self.buffer[start_index..full_item_len].to_string();
            self.buffer = self.buffer[full_item_len..].to_string();

            match parse_item_xml(&item_xml) {
                Ok((code, value)) => {
                    if !TRANSPORT_CODES.contains(&code.as_str())
                        && !METADATA_CODES.contains(&code.as_str())
                    {
                        continue;
                    }

                    if let Some(state) = transport_state(&code) {
                        self.session_manager
                            .update_transport_state("AirPlay", state)
                            .await;
                    } else if code == "mdst" {
                        self.staged_track_info.clear();
                    } else if let Some(value) = value {
                        self.staged_track_info.insert(code, value);
                        if self.staged_track_info.contains_key("minm")
                            && self.staged_track_info.contains_key("asar")
                        {
                            should_commit_metadata = true;
                        }
                    }
                }
                Err(err) => {
                    println!("SHAIRPORT: Failed to parse item block: {err:#}");
                    println!("SHAIRPORT: Malformed XML chunk was: {item_xml}");
                }
            }
        }

        if should_commit_metadata {
            println!("SHAIRPORT: Title and artist received. Committing metadata.");
            let artist = text_value(self.staged_track_info.get("asar"));
            let title = text_value(self.staged_track_info.get("minm"));
            let songid = format!("airplay-{artist}-{title}");
            let cover_url = self
                .staged_track_info
                .get("PICT")
                .and_then(|data| save_pict_data(&self.config.art_staging_dir, &songid, data).ok());

            self.session_manager
                .update_metadata(
                    "AirPlay",
                    Metadata {
                        songid: Some(songid),
                        title: Some(title),
                        artist: Some(artist),
                        album: None,
                        cover_url,
                    },
                )
                .await;
        }
    }
}

fn parse_item_xml(item_xml: &str) -> Result<(String, Option<Vec<u8>>)> {
    let doc = roxmltree::Document::parse(item_xml)?;
    let code_hex = doc
        .descendants()
        .find(|node| node.has_tag_name("code"))
        .and_then(|node| node.text())
        .unwrap_or_default();
    let code_bytes = decode_hex(code_hex)?;
    let code = String::from_utf8_lossy(&code_bytes).into_owned();

    let value = doc
        .descendants()
        .find(|node| node.has_tag_name("data"))
        .and_then(|node| {
            let content = node.text()?.trim();
            if content.is_empty() {
                None
            } else if node.attribute("encoding") == Some("base64") {
                STANDARD.decode(content).ok()
            } else {
                Some(content.as_bytes().to_vec())
            }
        });

    Ok((code, value))
}

fn decode_hex(input: &str) -> Result<Vec<u8>> {
    let mut out = Vec::with_capacity(input.len() / 2);
    for pair in input.as_bytes().chunks(2) {
        if pair.len() == 2 {
            let text = std::str::from_utf8(pair)?;
            out.push(u8::from_str_radix(text, 16)?);
        }
    }
    Ok(out)
}

fn transport_state(code: &str) -> Option<PlayerState> {
    match code {
        "prsm" => Some(PlayerState::Playing),
        "pend" => Some(PlayerState::Stopped),
        "paus" => Some(PlayerState::Paused),
        _ => None,
    }
}

fn save_pict_data(art_staging_dir: &PathBuf, songid: &str, image_bytes: &[u8]) -> Result<String> {
    fs::create_dir_all(art_staging_dir)?;
    let filename = format!("{:x}.tmp", md5::compute(songid.as_bytes()));
    let filepath = art_staging_dir.join(filename);
    if !filepath.exists() {
        fs::write(&filepath, image_bytes)?;
    }
    Ok(format!("file://{}", filepath.display()))
}

fn text_value(value: Option<&Vec<u8>>) -> String {
    value
        .map(|bytes| String::from_utf8_lossy(bytes).into_owned())
        .unwrap_or_default()
}
